### PR TITLE
Fix multiple CU and host application result mismatch issue

### PR DESCRIPTION
**Pull request by mamin506**
#### Problem solved by the commit
1. Host side CU sub device got incorrect XGQ response when multiple CUs are assigned to one XGQ.
2. All of the PL CUs and PS kernels is assigned to the same XGQ.
3. Coverity issue and unused variable

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Create local PS kernel test cases.

#### How problem was solved, alternative solutions (if any) and why they were rejected
1. When a CU check XGQ and if the response is not for this CU, cache the response.
2. Move xgqpdev to be the local variable in the loop
3. Initial info.cu_domain and remove ununsed variable 

#### Risks (if any) associated the changes in the commit
No

#### What has been tested and how, request additional testing if necessary
xbutil validate on vck5000 and u50
Some PS kernel test cases on vck5000

#### Documentation impact (if any)
No

### CODE REVIEW

**Review by jeffli-xilinx — APPROVED** [review]

**Comment by maxzhen** [comment]
@mamin506 , I may have missed something, let's talk offline.

**Comment by mamin506** [comment]
Hi @stsoe @rozumx , I guess you are added is because of I changed dkms.cmake to remove an unused source code cu_plram.c. It is for an experimental purpose CU type.

All of the other changes are inside KDS. You don't need to review this huge change unless you want to.

File: src/CMake/dkms.cmake [hunk 1]
```diff
@@ -290,7 +290,6 @@ SET (XRT_DKMS_COMMON_XRT_DRV
   common/drv/xrt_cu.c
```
Comment (rozumx) [inline src/CMake/dkms.cmake:290]: Copyright date (line 2) needs to be updated.

File: src/runtime_src/core/common/drv/cu_hls.c [hunk 2]
```diff
@@ -53,6 +68,18 @@ static inline void cu_write32(struct xrt_cu_hls *cu, u32 reg, u32 val)
 	iowrite32(val, cu->vaddr + reg);
 }
 
+static inline void cu_move_to_complete(struct xrt_cu_hls *cu, int status)
+{
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/cu_hls.c:73]: Unintialized pointer.

File: src/runtime_src/core/common/drv/cu_hls.c [hunk 3]
```diff
@@ -362,6 +392,70 @@ static bool cu_hls_reset_done(void *core)
...
+		return ret;
+
+	list_move_tail(&xcmd->list, &cu_hls->submitted);
+	return ret;
+}
+
+static struct kds_command *cu_hls_get_complete(void *core)
+{
+	struct xrt_cu_hls *cu_hls = core;
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/cu_hls.c:411]: Unitialized pointer.

File: src/runtime_src/core/common/drv/cu_hls.c [hunk 4]
```diff
@@ -362,6 +392,70 @@ static bool cu_hls_reset_done(void *core)
...
+	list_del_init(&xcmd->list);
+
+	return xcmd;
+}
+
+static int cu_hls_abort(void *core, void *cond,
+			bool (*match)(struct kds_command *xcmd, void *cond))
+{
+	struct xrt_cu_hls *cu_hls = core;
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/cu_hls.c:426]: Unitialized pointers.

File: src/runtime_src/core/common/drv/cu_hls.c [hunk 5]
```diff
@@ -362,6 +392,70 @@ static bool cu_hls_reset_done(void *core)
...
+	struct kds_command *next;
+	int ret = -EBUSY;
+
+	if (cu_hls->sw_reset) {
+		int time = 10 * 1000 * 1000;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/cu_hls.c:431]: Magic numbers.  What do these values mean?

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 6]
```diff
@@ -48,6 +48,19 @@ static inline void cu_write32(struct xrt_cu_fa *cu, u32 reg, u32 val)
 	iowrite32(val, cu->vaddr + reg);
 }
 
+static inline void cu_move_to_complete(struct xrt_cu_fa *cu, int status)
+{
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/fast_adapter.c:53]: Uninitilaized pointer.

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 7]
```diff
@@ -127,6 +140,7 @@ static void cu_fa_check(void *core, struct xcu_status *status, bool force)
 	struct xrt_cu_fa *cu_fa = core;
 	u32 task_count;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/fast_adapter.c:141]: Unitialized value.

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 8]
```diff
@@ -127,6 +140,7 @@ static void cu_fa_check(void *core, struct xcu_status *status, bool force)
...
 	u32 done = 0;
+	int i;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/fast_adapter.c:143]: Uninitalized value.

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 9]
```diff
@@ -190,6 +206,52 @@ static bool cu_fa_reset_done(void *core)
...
+		return ret;
+
+	list_move_tail(&xcmd->list, &cu_fa->submitted);
+	return ret;
+}
+
+static struct kds_command *cu_fa_get_complete(void *core)
+{
+	struct xrt_cu_fa *cu_fa = core;
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/fast_adapter.c:225]: Uninitialized pointer.

File: src/runtime_src/core/common/drv/fast_adapter.c [hunk 10]
```diff
@@ -190,6 +206,52 @@ static bool cu_fa_reset_done(void *core)
...
+	list_del_init(&xcmd->list);
+
+	return xcmd;
+}
+
+static int cu_fa_abort(void *core, void *cond,
+		       bool (*match)(struct kds_command *xcmd, void *cond))
+{
+	struct xrt_cu_fa *cu_fa = core;
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/fast_adapter.c:240]: Unintialized pointers.

**Review by maxzhen — APPROVED** [review]

File: src/runtime_src/core/common/drv/kds_core.c [hunk 11]
```diff
@@ -843,6 +844,29 @@ kds_del_scu_context(struct kds_sched *kds, struct kds_client *client,
 	if (submitted == completed)
 		goto skip;
 
+	if (kds->xgq_enable) {
+		int wait_ms;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/kds_core.c:848]: Unitialized variable.
```suggestion
		int wait_ms = 500;
```

File: src/runtime_src/core/common/drv/kds_core.c [hunk 12]
```diff
@@ -843,6 +844,29 @@ kds_del_scu_context(struct kds_sched *kds, struct kds_client *client,
...
+
+		wait_ms = 500;
```
Comment (rozumx) [inline src/runtime_src/core/common/drv/kds_core.c:850]:
```suggestion
```

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/cu_xgq.c [hunk 13]
```diff
@@ -104,6 +85,34 @@ static bool cu_xgq_reset_done(void *core)
 	return true;
 }
 
+static int cu_xgq_submit_config(void *core, struct kds_command *xcmd)
+{
+	struct xrt_cu_xgq *cu_xgq = core;
+	struct xgq_cmd_sq_hdr *hdr;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/cu_xgq.c:91]: Uninitialized variables.

File: src/runtime_src/core/pcie/driver/linux/xocl/userpf/Makefile [hunk 14]
```diff
@@ -44,7 +44,6 @@ drv_common-y   := $(common_dir)/kds_core.o \
 		  $(common_dir)/xrt_cu.o \
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/userpf/Makefile:44]: Copyright date line 2 needs to be updated.

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c [hunk 15]
```diff
@@ -20,12 +20,18 @@
 
 #define CLIENT_ID_BITS 7
 #define MAX_CLIENTS (1 << CLIENT_ID_BITS)
+#define CLIENT_ID_MASK (MAX_CLIENTS - 1)
 
 static uint16_t xocl_xgq_cid;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:25]: Uninitalized variable.

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c [hunk 16]
```diff
@@ -108,33 +117,113 @@ void xocl_xgq_notify(void *xgq_handle)
 	xocl_xgq_trigger_sq_intr(xgq);
 }
 
-int xocl_xgq_get_response(void *xgq_handle, int id, struct xgq_com_queue_entry *resp)
+static int xocl_xgq_handle_resp(struct xocl_xgq *xgq, int id, u64 resp_addr)
+{
+	struct xgq_com_queue_entry *resp = (struct xgq_com_queue_entry *)resp_addr;
+	struct xocl_xgq_client *client = &xgq->xx_clients[id];
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:124]: Unintialized pointer.

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c [hunk 17]
```diff
@@ -108,33 +117,113 @@ void xocl_xgq_notify(void *xgq_handle)
...
 	spin_unlock_irqrestore(&xgq->xx_lock, flags);
 	return ret;
 }
 
-int xocl_xgq_attach(void *xgq_handle, void *client, int *client_id)
+struct kds_command *xocl_xgq_get_command(void *xgq_handle, int id)
+{
+	struct xocl_xgq *xgq = (struct xocl_xgq *)xgq_handle;
+	struct xocl_xgq_client *client = &xgq->xx_clients[id];
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:182]: Unintialized pointer.

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c [hunk 18]
```diff
@@ -108,33 +117,113 @@ void xocl_xgq_notify(void *xgq_handle)
...
+
+	return xcmd;
+}
+
+int xocl_xgq_abort(void *xgq_handle, int id, void *cond,
+		   bool (*match)(struct kds_command *xcmd, void *cond))
+{
+	struct xocl_xgq *xgq = (struct xocl_xgq *)xgq_handle;
+	struct xocl_xgq_client *client = &xgq->xx_clients[id];
+	struct kds_command *xcmd;
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:204]: Unintialized pointers.

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c [hunk 19]
```diff
@@ -108,33 +117,113 @@ void xocl_xgq_notify(void *xgq_handle)
...
+	struct kds_command *next;
+	unsigned long flags = 0;
+	int ret = 0;
+
+	spin_lock_irqsave(&client->xxc_lock, flags);
+	list_for_each_entry_safe(xcmd, next, &client->xxc_submitted, list) {
+		if (!match(xcmd, cond))
+			continue;
+
+		/* TODO: Send abort XGQ command to device. */
```
Comment (rozumx) [inline src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:214]: Is this a place holder or is there missing functionality here.

**Review by rozumx — CHANGES_REQUESTED** [review]

File: src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c (hunk 15 above)
Comment (mamin506) [reply src/runtime_src/core/pcie/driver/linux/xocl/xocl_xgq.c:25]: Global variables are automatically initialized to 0 at the time of declaration. This is define in the C standard.

**Review by rozumx — APPROVED** [review]
Approved.  Thank you for making the changes.

**Review by keryell — COMMENTED** [review]
Just looking at this lately.
Avoid committing uncommented code. :-(

**Pull request merged**